jobs: improve job adoption

ajwerner opened the pull request:
#### jobs: don't hold mutex during adoption, launch in parallel

#### jobs: break up new stages of job lifecycle movement

In the PR which adopted the sqlliveness sessions, we shoved all of the stages
of adopting jobs into the same stage and we invoked that stage on each adoption
interval and on each sent to the adoption channel.

These stages are:

 * Cancel jobs
 * Serve pause and cancel requests
 * Delete claims due to dead sessions
 * Claim jobs
 * Process claimed jobs

This is problematic for tests which send on the adoption channel at a high
rate. One important thing to note is that all jobs which are sent on the
adoption channel are already claimed.

After this PR we move the first three steps above into the cancellation
loop we were already running. We also increase the default interval for
that loop as it was exceedingly frequent at 1s for no obvious reason.

Much of the testing changes are due to this cancelation loop duration
change. The tests in this package now run 3x faster (10s vs 30s).

Then, upon sends on the adoption channel, we just process claimed jobs.
When the adoption interval rolls around, then we attempt to both claim
and process jobs.

Release justification: bug fixes and low-risk updates to new functionality
Release note: None

Commit:
jobs: don't hold mutex during adoption, launch in parallel

Release justification: bug fixes and low-risk updates to new functionality
Release note: None

Commit:
jobs: break up new stages of job lifecycle movement

In the PR which adopted the sqlliveness sessions, we shoved all of the stages
of adopting jobs into the same stage and we invoked that stage on each adoption
interval and on each sent to the adoption channel.

These stages are:

 * Cancel jobs
 * Serve pause and cancel requests
 * Delete claims due to dead sessions
 * Claim jobs
 * Process claimed jobs

This is problematic for tests which send on the adoption channel at a high
rate. One important thing to note is that all jobs which are sent on the
adoption channel are already claimed.

After this PR we move the first three steps above into the cancellation
loop we were already running. We also increase the default interval for
that loop as it was exceedingly frequent at 1s for no obvious reason.

Much of the testing changes are due to this cancelation loop duration
change. The tests in this package now run 3x faster (10s vs 30s).

Then, upon sends on the adoption channel, we just process claimed jobs.
When the adoption interval rolls around, then we attempt to both claim
and process jobs.

Release justification: bug fixes and low-risk updates to new functionality
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/53589)
<!-- Reviewable:end -->

Commit:
jobs: minor tweaks increase adoption concurrency

This commit turns out to make a reasonably big difference.

Release justification: bug fixes and low-risk updates to new functionality
Release note: None

ajwerner commented:
I've `roachprod-stressrace`d  this for 16 minutes so far.

ajwerner commented:
On master the `typeorm` test takes:

```
--- PASS: typeorm (2425.84s)
--- PASS: typeorm (2510.47s)
--- PASS: typeorm (2554.91s)
PASS
```

With this PR it is:
```
--- PASS: typeorm (1774.17s)
--- PASS: typeorm (1786.25s)
--- PASS: typeorm (1821.85s)
PASS
```

Better but not amazing. cc @rafiss

ajwerner commented:
Actually setting the GC TTL and the merge queue setting gets us to:

```
--- PASS: typeorm (1444.19s)
--- PASS: typeorm (1463.58s)
--- PASS: typeorm (1530.58s)
```

The CPU profiles are rather hilarious still. I'm working on an RFC to deal with the root cause of that pain. I'll post the basic finding which I should have very soon.

Commit:
jobs: run GC at least as frequently as things expire

Release justification: low risk, high benefit changes to existing functionality
Release note: None

ajwerner commented:
A couple more tweaks to the KV layer gets us to:

```
=== RUN   typeorm
=== RUN   typeorm
=== RUN   typeorm
--- PASS: typeorm (1094.87s)
--- PASS: typeorm (1094.73s)
--- PASS: typeorm (1084.94s)
```

ajwerner commented:
In sum this + one of https://github.com/cockroachdb/cockroach/pull/53605 or https://github.com/cockroachdb/cockroach/pull/53603 + https://github.com/cockroachdb/cockroach/pull/53606 gets us to:

```
--- PASS: typeorm (1094.87s)
--- PASS: typeorm (1094.73s)
--- PASS: typeorm (1084.94s)
```

spaskob commented on pkg/jobs/registry.go:
```diff
@@ -298,15 +298,17 @@ func (r *Registry) Run(ctx context.Context, ex sqlutil.InternalExecutor, jobs []
-			return stop.ErrUnavailable
-		case <-ctx.Done():
-			return ctx.Err()
+		if !usingSQLLiveness || i == 0 {
```
please add a comment explaining the if condition

ajwerner replied:
Done.

ajwerner commented on pkg/jobs/registry.go:
I'm not sure I follow this comment. This loop was both the old and the new logic. I've now separated them.

spaskob replied:
The new logic is in `claimAndProcessJobs`

spaskob reviewed (commented):
PTAL
ping me directly if it is easier

spaskob reviewed (approved):
> Reviewed 12 of 15 files at r2, 2 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53589)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, and @spaskob)

___
*[pkg/jobs/adopt.go, line 77 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/53589#-MG4QiS99xYPVa0DAxmg:-MG4QiSA5RsdtWbRVg21:b-xkrmm) ([raw file](https://github.com/cockroachdb/cockroach/blob/df0e881727934d5df6020aa0547215efdfd6b81d/pkg/jobs/adopt.go#L77)):*
> ```Go
> // resumeClaimedJobs invokes r.resumeJob for each job in claimedToResume. It
> // does so concurrently.
> func (r *Registry) resumeClaimedJobs(
> ```

an observation: the slow bit of this operation is fetching the payload of the job to be run from the jobs table; an alternative approach may be to fetch all these in one query. I am not sure if this is better, just food for thought. 


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
TFTR!

bors r=spaskob

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53589#-:-MG4VHOm6tRU8BHGAlcP:bjggt9a)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @spaskob)

___
*[pkg/jobs/adopt.go, line 77 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/53589#-MG4QiS99xYPVa0DAxmg:-MG4UeaA9wX4Al-u38de:bodlyam) ([raw file](https://github.com/cockroachdb/cockroach/blob/df0e881727934d5df6020aa0547215efdfd6b81d/pkg/jobs/adopt.go#L77)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

an observation: the slow bit of this operation is fetching the payload of the job to be run from the jobs table; an alternative approach may be to fetch all these in one query. I am not sure if this is better, just food for thought. 
</blockquote></details>

In a world where the interfaces were different, I think you're right. I'd love to inject the set of currently running jobs under the sql query via a builtin or virtual table and then pull the job payload directly but that feels more disruptive at this point. I think this will work nicely for this release. 
___
*[pkg/jobs/registry.go, line 594 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53589#-MFn7kgXeplE0zdpgT7W-r2-594:-MG4VG3p7WkRHoa4fazd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1281a3c1085e74f566a7ea734492280e9b2aa10e/pkg/jobs/registry.go#L594)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

The new logic is in `claimAndProcessJobs`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajwerner commented:
bors r+

otan commented:
is this causing `TestTruncateCompletion` to fail in https://teamcity.cockroachdb.com/viewLog.html?buildId=2236228&buildTypeId=Cockroach_UnitTests?

ajwerner commented:
Maybe. 

bors r-

while I investigate

ajwerner commented:
bors r+

rafiss commented:
touches #52556
